src/main/java/org/spongepowered/api/util/command/args/GenericArguments.java

```diff
@@ -1247,4 +1254,90 @@ public Text getUsage(CommandSource src) {
             return this.element.getUsage(src);
         }
     }
+    
+    public static CommandElement entity(Text key, Game game) {
+        return new EntityCommandElement(key, game, false);
+    }
+    
+    public static CommandElement entityOrSource(Text key, Game game) {
+        return new EntityCommandElement(key, game, true);
+    }
+
+    private static class EntityCommandElement extends SelectorCommandElement {
+        private final Game game;
+        private final boolean returnSource;
+    
+        protected EntityCommandElement(Text key, Game game, boolean returnSource) {
+            super(key);
+            this.game = game;
+            this.returnSource = returnSource;
+        }
+    
+        @Override
+        protected Object parseValue(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            if (!args.hasNext() && this.returnSource) {
+                return tryReturnSource(source, args);
+            }
+    
+            Object state = args.getState();
+            try {
+                return super.parseValue(source, args);
+            } catch (ArgumentParseException ex) {
+                if (this.returnSource) {
+                    args.setState(state);
+                    return tryReturnSource(source, args);
+                } else {
+                    throw ex;
+                }
+            }
+        }
+    
+        @Override
+        protected Iterable<String> getChoices(CommandSource source) {
+            Set<Iterable<Entity>> worldEntities = new HashSet<Iterable<Entity>>();
+            for (World world : this.game.getServer().getWorlds()) {
+                worldEntities.add(world.getEntities());
+            }
+            return Iterables.transform(Iterables.concat(worldEntities), new Function<Entity, String>() {
+                @Nullable
+                @Override
+                public String apply(@Nullable Entity input) {
+                    if (input == null) {
+                        return null;
+                    }
+                    if (input instanceof Player) {
+                        return ((Player)input).getName();
+                    }
+                    return input.getUniqueId().toString();
+                }
+            });
+        }
+    
+        @Override
+        protected Object getValue(String choice) throws IllegalArgumentException {
+            UUID uuid = UUID.fromString(choice);
+            for (World world : this.game.getServer().getWorlds()) {
+                Optional<Entity> ret = world.getEntity(uuid);
+                if (ret.isPresent()) {
+                    return ret.get();
+                }
+            }
+            throw new IllegalArgumentException("Input value " + choice + " was not an entity");
+        }
+    
+        private Entity tryReturnSource(CommandSource source, CommandArgs args) throws ArgumentParseException {
+            // TODO get entity from ProxySource?
```

Comment:
How about we add a new interface, `ProxySourceTarget`, that is extended by `Entity` and `CommandSource`? That should take care of it.